includes/REST_API/Link_Controller.php

```diff
@@ -0,0 +1,255 @@
+ *
+ * Class Link_Controller
+ */
+class Link_Controller extends WP_REST_Controller {
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'web-stories/v1';
+		$this->rest_base = 'link';
+	}
+
+	/**
+	 * Registers routes for links.
+	 *
+	 * @see register_rest_route()
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'parse_link' ],
+					'permission_callback' => [ $this, 'parse_link_permissions_check' ],
+					'args'                => [
+						'url'       => [
+							'description'       => __( 'The URL to process.', 'web-stories' ),
+							'type'              => 'string',
+							'required'          => true,
+							'sanitize_callback' => 'esc_url_raw',
+						],
+					],
+				],
+			]
+		);
+	}
+
+	/**
+	 * Parses a URL to return some metadata for inserting links.
+	 *
+	 * @param WP_REST_Request $request Full data about the request.
+	 *
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function parse_link( $request ) {
+		$url = $request['url'];
+
+		/**
+		 * Filters the link data TTL value.
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 * @param string $url The attempted URL.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_link_data_cache_ttl', DAY_IN_SECONDS, $url );
+		$cache_key = 'web_stories_link_data_' . md5( $url );
+
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			return json_decode( $data, true );
+		}
+
+		$title       = '';
+		$image       = '';
+		$description = '';
+
+		$data = [
+			'title'       => $title,
+			'image'       => $image,
+			'description' => $description,
+		];
+
+		$response = wp_safe_remote_get(
+			$url,
+			[
+				'limit_response_size' => 153600, // 150 KB.
+			]
+		);
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), array( 'status' => 404 ) );
+		}
+
+		$html  = wp_remote_retrieve_body( $response );
+
+		// Strip <body>.
+		$html_head_end = stripos( $html, '</head>' );
+		if ( $html_head_end ) {
+			$html = substr( $html, 0, $html_head_end );
+		}
+
+		if ( ! $html ) {
+			set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+			return new WP_Error( 'rest_invalid_url', get_status_header_desc( 404 ), array( 'status' => 404 ) );
+		}
+
+		$doc                      = new DOMDocument();
+		$doc->strictErrorChecking = false;
+
+		// Suppress warnings generated by loadHTML.
+		$errors = libxml_use_internal_errors( true );
+		$doc->loadHTML( $html );
+		libxml_use_internal_errors( $errors );
+		$xpath = new DOMXpath( $doc );
+
+		// Link title.
```

Comment:
Could we perhaps split this function up a bit for readability to separate getting title, icon, description?

Reply:
Yes. I'll extract the logic into a separate method, similar to what Jonny mentioned above